src/MetaMorpho.sol

```diff
@@ -345,51 +345,52 @@ contract MetaMorpho is ERC4626, ERC20Permit, Ownable2Step, Multicall, IMetaMorph
-            MarketAllocation memory allocation = withdrawn[i];
+        for (uint256 i; i < allocations.length; ++i) {
+            MarketAllocation memory allocation = allocations[i];
             Id id = allocation.marketParams.id();
 
             if (allocation.marketParams.loanToken != asset()) revert ErrorsLib.InconsistentAsset(id);
 
-            // Guarantees that unknown frontrunning donations can be withdrawn, in order to disable a market.
-            uint256 shares;
-            if (allocation.assets == type(uint256).max) {
-                shares = MORPHO.supplyShares(id, address(this));
+            (uint256 totalSupplyAssets, uint256 totalSupplyShares,,) =
+                MORPHO.expectedMarketBalances(allocation.marketParams);
 
-                allocation.assets = 0;
-            }
+            uint256 supplyShares = MORPHO.supplyShares(id, address(this));
+            uint256 supplyAssets = supplyShares.toAssetsDown(totalSupplyAssets, totalSupplyShares);
+            uint256 withdrawn = supplyAssets.zeroFloorSub(allocation.assets);
 
-            (uint256 withdrawnAssets, uint256 withdrawnShares) =
-                MORPHO.withdraw(allocation.marketParams, allocation.assets, shares, address(this), address(this));
+            if (withdrawn > 0) {
+                // Guarantees that unknown frontrunning donations can be withdrawn, in order to disable a market.
+                uint256 shares;
+                if (allocation.assets == 0) {
+                    shares = supplyShares;
+                    withdrawn = 0;
+                }
 
-            totalWithdrawn += withdrawnAssets;
+                (uint256 withdrawnAssets, uint256 withdrawnShares) =
+                    MORPHO.withdraw(allocation.marketParams, withdrawn, shares, address(this), address(this));
 
-            emit EventsLib.ReallocateWithdraw(_msgSender(), id, withdrawnAssets, withdrawnShares);
-        }
+                emit EventsLib.ReallocateWithdraw(_msgSender(), id, withdrawnAssets, withdrawnShares);
 
-        uint256 totalSupplied;
-        for (uint256 i; i < supplied.length; ++i) {
-            MarketAllocation memory allocation = supplied[i];
-            Id id = allocation.marketParams.id();
-            uint256 supplyCap = config[id].cap;
+                totalWithdrawn += withdrawnAssets;
+            } else {
+                uint256 supplied = allocation.assets.zeroFloorSub(supplyAssets);
+                if (supplied == 0) continue;
 
-            if (supplyCap == 0) revert ErrorsLib.UnauthorizedMarket(id);
+                uint256 supplyCap = config[id].cap;
+                if (supplyCap == 0) revert ErrorsLib.UnauthorizedMarket(id);
 
-            (uint256 suppliedAssets, uint256 suppliedShares) =
-                MORPHO.supply(allocation.marketParams, allocation.assets, 0, address(this), hex"");
+                if (supplyAssets + supplied > supplyCap) revert ErrorsLib.SupplyCapExceeded(id);
```

Comment:
It's minor but this does not compute exactly the same thing as if the check was done after the `supply` call: the variable `supplyAssets` is computed using old values of the total supply and the total supply shares

Reply:
Why would they be "old"? Old compared to what? These values are not expected to change in this execution context right?